dap-mode.el

```diff
@@ -1167,25 +1186,38 @@ after selecting configuration template."
                            (error "There is no debug provider for language %s"
                                   (or (plist-get debug-args :type) "'Not specified'")))))
 
-(defun dap-debug-edit-template (debug-args)
-  "Edit template DEBUG-ARGS."
-  (interactive (list (dap--select-template)))
-  (with-current-buffer (or (get-buffer "*DAP Templates*")
-                           (with-current-buffer (get-buffer-create "*DAP Templates*")
-                             (emacs-lisp-mode)
-                             (current-buffer)))
-    (goto-char (point-max))
-    (insert "\n\n(dap-debug `(")
-    (-let ((column (current-column))
-           ((fst snd . rst) debug-args))
-      (insert (format "%s %s" fst (prin1-to-string snd)))
-      (cl-loop for (k v) on rst by (function cddr)
-               do (progn
-                    (insert "\n")
-                    (--dotimes column (insert " "))
-                    (insert (format "%s %s" k (prin1-to-string v))))))
-    (insert "))"))
-  (pop-to-buffer "*DAP Templates*"))
+(defun dap-debug-edit-template (&optional parg debug-args)
+  "Edit registered template DEBUG-ARGS.
+When being invoked with prefix argument, poping up the prepopulated version of the template.
+Otherwise, return its original version. After registration, the new template can be used
+normally with dap-debug"
+
+  (interactive "P")
+  (let ((debug-args (dap--select-template(not parg))))
+    (progn
+      (message "debug-args: %s" debug-args)
+      (with-current-buffer (or (get-buffer "*DAP Templates*")
+			       (with-current-buffer (get-buffer-create "*DAP Templates*")
+				 (emacs-lisp-mode)
+				 (current-buffer)))
+	  (goto-char (point-max))
+	  (insert
+	   (format "\n\n(dap-register-debug-template \"%s%s\"\n"
```

Comment:
I will have to think more about that register call, it will slightly diverge from what VScode does. I think that something similar what suggested by @danielmartin . We need to decide also how we are going to persist the templaces, e. g. we may do that in `workspace-folder/.dap-templates` or we can support the vscode format.

Reply:
We could support the `launch.json` format, yes. Interoperability with VS Code is a good thing.

Reply:
Yes, I completely agree with @yyoncho that we should differentiate between the template and the configuration. This is also the reason why I though about a new command for edit/create configuration based on given template registered by each debug provider. But as @yyoncho reminder, new command should be avoided.

@yyoncho @danielmartin how about re-using lsp-workspace-root, we can have dedicated file for each project by hashing the project path and concatenating it with project name. Then putting the file under our dap-configurations folder. Besides, I'm more preferred to put this under the same location with our .dap-breakpoints since we will not affect the workspace repo.
Similar modification should be done for dap-breakpoints as well. Breakpoints for a project should not be visible inside another project view.

Reply:
> Besides, I'm more preferred to put this under the same location with our .dap-breakpoints since we will not affect the workspace repo.

We can make that configurable via variable, so we will support both options(we might calculate the storage path relatively to the project root which will work for full paths like /home/user/.storage/ and .storage). I do not like the idea about hashing since ideally the file should be human readable and the hashing is breaking this. 

> Similar modification should be done for dap-breakpoints as well. Breakpoints for a project should not be visible inside another project view.

You might have 2 project and one of them is referencing the other, in this case you want to hit the breakpoint from projectA even if you have started the debug process from projectB. This is  common for java projects.

Reply:
> We can make that configurable via variable, so we will support both options

Agree.
> the file should be human readable and the hashing is breaking this.

We still have the project name there, the hash is just only a short stuff to distinguish projects that have the same name but different location. Otherwise, we may need to have some other thing.
>You might have 2 project and one of them is referencing the other,

With two different breakpoint files, one can also references to files that belong to other project with a normal absolute path, right?

Reply:
> We still have the project name there, the hash is just only a short stuff to distinguish projects that have the same name but different location. Otherwise, we may need to have some other thing.

I was thinking about not having the project name at all. If you want per project configuration - use the dir local to redirect the configuration into custom file. By default, keep then in a simple list globally.  

> With two different breakpoint files, one can also references to files that belong to other project with a normal absolute path, right?

You can make it work somehow but this will require adding a method which is - "add breakpoint and ask for project to use" since you will have both projects active and dap-mode won't know which project the breakpoint is for.